bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java

```diff
@@ -119,73 +119,76 @@ public void activate() {
-                        resources.addAll(collectResources(f));
-                    }
-                }
-                resources.forEach(this::importFileWhenReady);
-            }
+    protected Optional<String> getScriptType(Path scriptFilePath) {
+        String fileName = scriptFilePath.toString();
+        int index = fileName.lastIndexOf(".");
+        if (index == -1) {
+            return Optional.empty();
         }
+        String fileExtension = fileName.substring(index + 1);
+
+        // ignore known file extensions for "temp" files
+        if (EXCLUDED_FILE_EXTENSIONS.contains(fileExtension) || fileExtension.endsWith("~")) {
+            return Optional.empty();
+        }
+        return Optional.of(fileExtension);
     }
 
     /**
-     * Collects all resources from the specified file or directory,
-     * possibly including subdirectories.
-     *
-     * The results will be sorted.
+     * Gets the individual start level for a given file
+     * <p />
+     * The start level is derived from the name and location of
+     * the file by {@link #START_LEVEL_PATTERNS}. If no match is found, {@link StartLevelService#STARTLEVEL_RULEENGINE}
+     * is used.
      *
-     * @param file the file or directory to import resources from
+     * @param scriptFilePath the {@link Path} to the script
+     * @return the start level for this script
      */
-    private Collection<ScriptFileReference> collectResources(File file) {
-        Collection<ScriptFileReference> resources = new TreeSet<>();
-        if (!file.exists()) {
-            return resources;
-        }
-
-        if (file.isDirectory()) {
-            if (watchSubDirectories()) {
-                File[] files = file.listFiles();
-                if (files != null) {
-                    for (File f : files) {
-                        if (!f.isHidden()) {
-                            resources.addAll(collectResources(f));
-                        }
-                    }
+    protected int getStartLevel(Path scriptFilePath) {
+        for (Pattern p : START_LEVEL_PATTERNS) {
+            Matcher m = p.matcher(scriptFilePath.toString());
+            if (m.find() && m.groupCount() > 0) {
+                try {
+                    return Integer.parseInt(m.group(1));
+                } catch (NumberFormatException nfe) {
+                    logger.warn("Extracted start level {} from {}, but it's not an integer. Ignoring.", m.group(1),
+                            scriptFilePath);
                 }
             }
-        } else {
-            try {
-                URL url = file.toURI().toURL();
-                resources.add(new ScriptFileReference(url));
-            } catch (MalformedURLException e) {
-                // can't happen for the 'file' protocol handler with a correctly formatted URI
-                logger.debug("Can't create a URL", e);
-            }
         }
-        return resources;
+
+        return StartLevelService.STARTLEVEL_RULEENGINE;
```

Comment:
@ccutrer @florian-h05 I wonder if this is really a good choice. Start level 40 is named "rules loaded", so shouldn't they be loaded at start level 30 instead after persistence has restored item states?

Reply:
This is a good question.
Looking at https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/runtime/services.cfg, start level 40 is that dsl rules and UI rules are loaded. (If I understand that right?)
So I would say that JSR223 file based rules belong to start level 40 too.

I guess we also have to look out, that the addons are loaded before their scripts are loaded by the `AbstractScriptFileWatcher`, but honestly I have no experience with the start levels and the start order in openHAB.

Reply:
The difficult thing to figure out is if all script file watchers have finished their initial script processing because we don't know which exist. 

If we move the "automation:scriptEngingeFactories" to start level 30, add a new interface `ScriptFileWatcher` with `void boolean initialLoadingCompleted` and set another ready marker "jsr223:initialLoadingCompleted" which is needed for start level 40.

Reply:
My ideal situation:

 * JSR223 scripts by default act as close to DSL rules as possible. Seems like loading at SL30 would mirror that.
 * a user can override the start level to be as early as possible. I have scripts that don't add rules, but extend openHAB such as adding items or profiles, or custom units (I should probably get that one added to core). So if I can get something to run as early as SL10, that would be ideal.

I haven't looked too close into how the ScriptFileWatcher action actually starts, or when automation addons start. Mostly in ensuring the ordering of the configured start levels for scripts is respected.

Reply:
@florian-h05 @ccutrer I have improved the code so that it handles startup in a more predictable way. There might be issues if files are modified prior to the initial import (i.e. before the system reached start level 30). 

But setting the import to start level 30 in conjunction with #3275 makes sure script engines are available before we import files. I also added a new interface that allows us to later set a ready marker for "all script file watchers finished their initial import", but I postponed implementation of that because it depends on #3275.

Reply:
Sounds good, thanks :+1: